### PR TITLE
Standalone recorder proxy

### PR DESCRIPTION
**Pull request by johnerikhalse**

### CODE REVIEW

File: recorderproxy/https.go [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+func httpError(w io.WriteCloser, ctx *recordContext, err error) {
+	if _, err := io.WriteString(w, "HTTP/1.1 502 Bad Gateway\r\n\r\n"); err != nil {
+		ctx.Warnf("Error responding to client: %s", err)
+	}
+	if err := w.Close(); err != nil {
+		ctx.Warnf("Error closing client connection: %s", err)
+	}
+}
+
+func (proxy *RecorderProxy) NewConnectDialToProxy(https_proxy string) func(network, addr string) (net.Conn, error) {
```
Comment (maeb) [inline recorderproxy/https.go]: https_proxy -> httpsProxy

File: recorderproxy/https.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+	return proxy.NewConnectDialToProxyWithHandler(https_proxy, nil)
+}
+
+func (proxy *RecorderProxy) NewConnectDialToProxyWithHandler(https_proxy string, connectReqHandler func(req *http.Request)) func(network, addr string) (net.Conn, error) {
```
Comment (maeb) [inline recorderproxy/https.go]: https_proxy -> httpsProxy

File: recorderproxy/https.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+				}
+			}
+			proxyClient.Close()
+
+			return
+		}
+		defer rawClientTls.Close()
+		clientTlsReader := bufio.NewReader(rawClientTls)
+		for !isEof(clientTlsReader) {
+			req, err := http.ReadRequest(clientTlsReader)
```
Comment (maeb) [inline recorderproxy/https.go]: Make the body of the loop a function to make it more readable and avoid `defer` in a loop.

File: main.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	timeout := viper.GetDuration("timeout")
+	conn := recorderproxy.NewConnections()
+	err = conn.Connect(viper.GetString("content-writer"), viper.GetString("dns-resolver"), viper.GetString("browser-controller"), timeout)
+	if err != nil {
+		log.Fatalf("Could not connect to services: %v", err)
+	}
+
+	fmt.Printf("Using cache at %s\n", viper.GetString("cache"))
```
Comment (maeb) [inline main.go]: `cache` is unused in main function. Should be used as an argument to NewRecorderProxy.

File: recorderproxy/recorderproxy.go [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+	}
+
+	cache := viper.GetString("cache")
+	if cache != "" {
+		cu, _ := url.Parse("http://" + cache)
+		r.RoundTripper.Proxy = http.ProxyURL(cu)
+		r.ConnectDial = r.NewConnectDialToProxy(cache)
+	}
+
+	r.Logger = log.New(os.Stderr, "", log.LstdFlags)
```
Comment (maeb) [inline recorderproxy/recorderproxy.go:101]: Should every recorderProxy create a new logger? Maybe a shared logger should be created in `main` function and be an argument to `NewRecorderProxy`?

File: recorderproxy/recorderproxy.go [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+		addr:              ":" + strconv.Itoa(port),
+		ConnectionTimeout: connectionTimeout,
+
+		NonproxyHandler: http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			http.Error(w, "This is a proxy server. Does not respond to non-proxy requests.", 500)
+		}),
+		RoundTripper: NewRpRoundTripper(),
+	}
+
+	cache := viper.GetString("cache")
```
Comment (maeb) [inline recorderproxy/recorderproxy.go]: I suggest sending cache as a parameter instead of using viper.

File: recorderproxy/https.go [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+			if _, err = io.WriteString(rawClientTls, "\r\n"); err != nil {
+				ctx.Warnf("Cannot write TLS response chunked trailer from mitm'd client: %v", err)
+				return
+			}
+		}
+		ctx.Logf("Exiting on EOF")
+	}()
+}
+
+func httpError(w io.WriteCloser, ctx *recordContext, err error) {
```
Comment (maeb) [inline recorderproxy/https.go:226]: `err` argument is unused in function body. Isn't the point of the function to handle the error in some way?

File: recorderproxy/recordcontext.go [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+	}
+
+	return ctx
+}
+
+func (ctx *recordContext) init(req *http.Request) *recordContext {
+	cwcCtx, cwcCancel := context.WithTimeout(context.Background(), ctx.proxy.ConnectionTimeout)
+	cwc, err := ctx.proxy.conn.ContentWriterClient().Write(cwcCtx)
+	if err != nil {
+		log.Fatalf("Error connecting to content writer, cause: %v", err)
```
Comment (maeb) [inline recorderproxy/recordcontext.go]: `log.Fatalf` is equivalent to `log.Printf` followed by `os.Exit(1)`. So code following this statement does not execute.

File: recorderproxy/recordcontext.go [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+		cwcCancel()
+		return nil
+	}
+
+	bccCtx, bccCancel := context.WithTimeout(context.Background(), ctx.proxy.ConnectionTimeout)
+	bcc, err := ctx.proxy.conn.BrowserControllerClient().Do(bccCtx)
+	if err != nil {
+		log.Fatalf("Error connecting to browser controller, cause: %v", err)
```
Comment (maeb) [inline recorderproxy/recordcontext.go]: log.Fatalf is equivalent to log.Printf followed by os.Exit(1). So code following this statement does not execute.

File: recorderproxy/connection.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+	opts = append(opts, grpc.WithInsecure(), grpc.WithBlock())
+
+	dialCtx, dialCancel := context.WithTimeout(context.Background(), connectTimeout)
+	defer dialCancel()
+
+	// Set up ContentWriterClient
+	clientConn, err := grpc.DialContext(dialCtx, contentWriterAddr, opts...)
+	if err != nil {
+		log.Fatalf("fail to dial contentwriter at %v: %v", contentWriterAddr, err)
```
Comment (maeb) [inline recorderproxy/connection.go]: log.Fatalf is equivalent to log.Printf followed by os.Exit(1). So code following this statement does not execute.

File: recorderproxy/connection.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+	}
+	c.contentWriterClientConn = clientConn
+	c.contentWriterClient = contentwriterV1.NewContentWriterClient(clientConn)
+
+	log.Printf("Connected to contentwriter")
+
+	// Set up DnsResolverClient
+	clientConn, err = grpc.DialContext(dialCtx, dnsResolverAddr, opts...)
+	if err != nil {
+		log.Fatalf("fail to dial dns resolver at %v: %v", dnsResolverAddr, err)
```
Comment (maeb) [inline recorderproxy/connection.go]: log.Fatalf is equivalent to log.Printf followed by os.Exit(1). So code following this statement does not execute.

File: recorderproxy/connection.go [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+	}
+	c.dnsResolverClientConn = clientConn
+	c.dnsResolverClient = dnsresolverV1.NewDnsResolverClient(clientConn)
+
+	log.Printf("Connected to dns resolver")
+
+	// Set up BrowserControllerClient
+	clientConn, err = grpc.DialContext(dialCtx, browserControllerAddr, opts...)
+	if err != nil {
+		log.Fatalf("fail to dial browser controller at %v: %v", browserControllerAddr, err)
```
Comment (maeb) [inline recorderproxy/connection.go]: log.Fatalf is equivalent to log.Printf followed by os.Exit(1). So code following this statement does not execute.

File: recorderproxy/recorderproxy.go [hunk 13]
```diff
@@ -0,0 +1,448 @@
...
+	//   be communicated by proxies over further connections.
+	r.Header.Del("Connection")
+}
+
+// Standard net/http function. Shouldn't be used directly, http.Serve will use it.
+func (proxy *RecorderProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	//r.Header["X-Forwarded-For"] = w.RemoteAddr()
+	if r.Method == "CONNECT" {
+		proxy.handleHttps(w, r)
+	} else {
```
Comment (maeb) [inline recorderproxy/recorderproxy.go:399]: Should there be a `handleHttp(w, r)` method symmetrical to `handleHttps(w,r)`?

**Review by maeb — CHANGES_REQUESTED** [review]
Logging is not consistent. It is a mix of `fmt.Printxxx`, `log.Printxxx`, new instances of `log`. Each recorderProxy creates it's own log instance. In addition the recorderContext uses it's own logger interface and helper functions using the global logger. It is confusing.

Errors is not handled consistently. Either the errors should explicitly be ignored using underscore (and maybe a comment) or handled.

File: recorderproxy/recorderproxy.go (hunk 5 above)
Comment (johnerikhalse) [reply recorderproxy/recorderproxy.go:101]: Won't fix for now

**Pull request merged**